tools/ClassGen/NodeBuilder.h

```diff
@@ -229,6 +308,18 @@ class Builder {
     return NodeBuilder(hStream, cStream, dStream, name, isBackendSpecific);
   }
 
+  /// Declare a new enum and generate code for it.
+  EnumBuilder *newEnum(const std::string &name) {
```

Comment:
I made this return a pointer because my understanding of r-values and move semantics is not strong enough for me to express this in those terms. If I make this return an `EnumBuilder`, the destructor gets called twice (once for for destroying the copy made by this function, another time for destroying the copy that resides in the caller's scope) which causes the enum definition to be emitted twice.

Some help expressing this without pointers would be awesome.

Reply:
I'm not sure I understand your comment exactly but if you return an `EnumBuilder` from this function directly it should be no problem and then use an rvalue reference for addMember like `NodeBuilder &addMember(EnumBuilder &&eb, const std::string &name)` (and also move into `enumMembers_ ` inside of addMember) it should fix the problem.

Reply:
This will only allow an enum to be used in one node though so it may not be what you want.

Reply:
Yeah, one of the goals of this change is to allow multiple nodes to have a member of the same type of enum, so this solution is not going to work.

Reply:
I kinda feel like this is a symptom of an overly-clever use of the destructor to emit code... I wouldn't mind an explicit `emit()` method.